doc/design/model_zoo_dev.md

```diff
@@ -0,0 +1,205 @@
+integrated into the corp data infrastructure.  Developers in these
+companies don't need to install any software, but can use the Web GUI
+of their data infrastructure.
+
+## Model Development in Python
+
+### Define a Model
+
+For clarify, let us consider a deep learning model definition is a
+Keras class in Python.  For example, the following class in
+`my_keras_model.py`:
+
+```python
+import sqlflow
+
+class MyKerasModel(tf.keras.Model):
+   def __init__(self, ...):
+      ...
+```
+
+For detailed instructions, please refer to this
+[document](https://github.com/sql-machine-learning/models/blob/develop/doc/customized%2Bmodel.md).
+
+### Test and Verify
+
+To verify that this model deifniton works, we can call Keras
+high-level API like `fit` to train a model and then evaluate it.
+
+After that, we want to verify it works with SQLFlow.  This depends on
+the SQLFlow high-level Python API, which is a Python abstraction of
+the SQL syntax extension introduced by SQLFlow.
+
+### SQLFlow High-level API
+
+For example, in SQLFlow we can write
+
+```sql
+SELECT ... TO TRAIN model_def WITH ... COLUMNS ... INTO trained_model;
+```
+
+Accordingly, in Python, we can write
+
+```python
+import sqlflow
+sqlflow.train("SELECT ...",
+              model_def,
+              with={"param1":value, "param2":value2},
+			  columns={"column1":lambda(x):..., "column2":labmda(x):...},
+			  into="trained_model")
+```
+
+### Unit Tests
+
+With the high-level API, we can test the model definition by writing a
+unit test file like `test_my_keras_model.py`.
+
+```python
+import sqlflow
+
+def test_MyKerasModel():
+  ## 初始化。dbhost：本地数据库地址；AI引擎默认是本地 SQLFlow server 中的TensorFlow, XGBoost等
+  sqlflow.init(dbhost="mysql://127.0.0.1:3306")
+  sqlflow.train("SELECT ...", MyKerasModel, ... into="my_trained_model")
+  auc, f1 = sqlflow.evaluate("SELECT ...", using="my_trained_model", metric=["AUC", "F1"])
+  unittest.assert(auc > 0.9)
+  unittest.assert(f1 > 0.5)
+```
+
+The above exmaple connects to a MySQL database system.  We can connect
+to other supported database systems as well.
+
+```python
+import sqlflow
+
+def test_MyKerasModel():
+  ## 初始化SQLFlow。 project： maxcompute project；AI引擎默认是PAI
+  sqlflow.init(dbhost="maxcompute://[AK]:[SK]@service-corp.odps.aliyun-inc.com/api?curr_project=alifin_jtest_dev&scheme=http")
+  sqlflow.train("SELECT ...", MyKerasModel, ... into="my_trained_model")
+  auc, f1 = sqlflow.evaluate("SELECT ...", using="my_trained_model", metric=["AUC", "F1"])
+  unittest.assert(auc > 0.9)
+  unittest.assert(f1 > 0.5)
+```
+
+We can run these unit tests by typing the following command.
+
+```bash
+pytest
+```
+
+### Test Using SQLFlow Command-line Tool
+
+Like MySQL has a command-line client `mysql`, we provide `sqlflow`.
+Using this command-line tool, we can run the following command to use
+the model definition.
+
+```bash
+$ sqlflow
+sqlflow > SELECT * from iris
+TO TRAIN MyKerasModel
+LABEL class
+INTO my_iris_model;
+```
+
+## Submit Models
```

Comment:
What does it mean to "submit models"? Does it mean to submit the Python source code including the model definition to GitHub/GitLab, or packing the Python source code into a Docker image and submit to a Docker registry, or submit a SQL statement that uses the model definition for the execution by the SQLFlow server?

Reply:
将 model definition 发布到某个 SQLFlow server，供这个 SQLFlow server 上的用户使用该模型。

Reply:
我们会说 docker push 也会说 git push。这些都是 Docker 和 Git 技术中的术语，有特定的意思。SQLFlow 里没有 push 这个术语。这里到底是想说什么呢？具体回答我上面的问题，应该可以帮助梳理清楚想法。

Reply:
Maybe `PUBLISH <model_def> <model_name>` is better, `model_def` is a folder, which inculdes `__init__.py`, `model_def.py` and `requirements.txt` at least. `PUBLISH` would execute the following steps:
- send `model_def` folder to the server 
- server execute `docker build && docker push` to a registry server, the image name is generated according to `model_name`

In advance, I suppose to add `VERIFY MODEL_DEF` to verify the model definition is correct.

Reply:
@Yancey1989 我不明白这个 submit model 到底是想做什么？因为不知道想做什么，所以我不理解你说的做法是否有道理。

Reply:
I think `MyKerasModel` should be related with a git repo. In this `submit model` step, we commit and push `MyKerasModel` and requirements into git and then trigger the docker build process automatically.
And then a new docker image is built and push into the docker registry service.

Reply:
@brightcoder01 Conceptually, packing a piece of software shouldn't rely on Git or SVN. Instead, we need dependency management like requirements.txt, Maven, and Dockerfile. If we plan to pack the model definition into a Docker image, then we need Docker.

@Yancey1989 I generally understand the purpose.  I prefer to run docker build and push from the shell instead of wrapping the shell commands in SQL statements.  It is almost impossible for us to propose such a SQL syntax extension to the standard committee.